artichoke-backend/src/extn/core/string/string_test.rb

```diff
@@ -47,6 +48,18 @@ def string_byteslice
   raise unless s.byteslice(5, 3) == 'fgh'
   raise unless s.byteslice(5, -10).nil?
   raise unless s.byteslice(5, -2).nil?
+  # range
+  raise unless s.byteslice(1..4) == 'bcde'
+  raise unless s.byteslice(1..-1) == 'bcdefghijk'
+  raise unless s.byteslice(10..-1) == 'k'
+  raise unless s.byteslice(20..-1).nil?
+  raise unless s.byteslice(20..-20).nil?
+  raise unless s.byteslice(2..-20) == ''
+  raise unless s.byteslice(-1..20) == 'k'
+  raise unless s.byteslice(-20..20).nil?
+  raise unless s.byteslice(-5..-1) == 'ghijk'
+  raise unless s.byteslice(-5..1) == ''
+  raise unless s.byteslice(-5..8) == 'ghi'
```

Comment:
I think I've figured out all the permutations of ranges that could be provided (and their `irb` values):

`s` =  start
`e` =  end
`S` = -start
`E` = -end
`x` =  start/end same value
`X` = -start/-end same value
* =  same element, different signs, see range

```plain
  "abcdefghijk".byteslice(range)
   |---------|   

# Normal ranges
   x---------|   =   0..0   | "a"
   se--------|   =   0..1   | "ab"
   s---------e   =   0..10  | "abcdefghijk"
   |s-------e|   =   1..9   | "bcdefghij"
   |--------se   =   9..10  | "jk"
   |--------s-e  =   9..11  | "jk"
   |---------x   =  10..10  | "k"
   |---------se  =  10..11  | "k"
   |---------|x  =  11..11  | ""
   |---------|se =  11..12  | ""
  
# Inverse ranges
   es--------|   =   1..0   | "" 
   e---------s   =  10..0   | "" 
   |e-------s|   =   9..1   | ""
   |--------es   =  10..9   | ""
   |--------e-s  =  11..9   | ""
   |---------es  =  11..10  | ""
   |---------|es =  12..11  | nil
  
# S-Negative normal ranges
  Se---------|   = -12..0   | nil 
  S-e--------|   = -12..1   | nil
   *---------|   = -11..0   | "a"
   Se--------|   = -11..1   | "ab"
   S---------e   = -11..10  | "abcdefghijk"
   S----------e  = -11..11  | "abcdefghijk"
   |S-------e|   = -10..9   | "bcdefghij"
   |--------Se   =  -2..10  | "jk"
   |---------*   =  -1..10  | "k"
  
# E-Negative normal ranges
   *---------|   =   0..-11 | "a"
   sE--------|   =   0..-10 | "ab"
   s---------E   =   0..-1  | "abcdefghijk"
   |s-------E|   =   1..-2  | "bcdefghij"
   |--------sE   =   9..-1  | "jk"
   |---------*   =  10..-1  | "k"

# S-Negative and E-Negative inverse ranges
  Es---------|   =   0..-12 | ""
  E-s--------|   =   1..-12 | ""
   Es--------|   =   1..-11 | ""
   |--------Es   =  10..-2  | ""
   |--------E-s  =  11..-2  | ""
   |---------Es  =  11..-1  | ""

# Dbl-Negative ranges
 SE|---------|   = -13..-12 | nil
  X|---------|   = -12..-12 | nil
  SE---------|   = -12..-11 | nil
  S-E--------|   = -12..-10 | nil
   X---------|   = -11..-11 | "a"
   SE--------|   = -11..-10 | "ab"
   S---------E   = -11..-1  | "abcdefghijk"
   |--------SE   =  -2..-1  | "jk"
   |---------X   =  -1..-1  | "k"

# Dbl-Negative inverse ranges
 ES|---------|   = -12..-13 | nil
  ES---------|   = -11..-12 | ""
  E-S--------|   = -10..-12 | ""
   ES--------|   = -10..-11 | ""
   E---------S   =  -1..-11 | ""
   |--------ES   =  -1..-2  | ""
```

I think there might be some cases that can be added if you're really keen :smile: